src/NHSD.BuyingCatalogue.Organisations.Api/Services/EmailAddress.cs

```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using MimeKit;
+
+namespace NHSD.BuyingCatalogue.Organisations.Api.Services
+{
+    /// <summary>
+    /// An e-mail address.
+    /// </summary>
+    internal sealed class EmailAddress
+    {
+        private string address;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EmailAddress"/> class.
+        /// </summary>
+        /// <remarks>Required by <see cref="System.Text.Json"/>.</remarks>
+        public EmailAddress()
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EmailAddress"/> class
+        /// with the given name and address.
+        /// </summary>
+        /// <param name="name">An optional </param>
+        /// <param name="address"></param>
+        internal EmailAddress(string name, string address)
```

Comment:
Would it be better to make this class immutable? I was under the impression that throwing an exception on the setting of the Address property was an anti pattern. It would not be obvious to a developer using this class that it throws. To reduce the surface area to get something wrong, would it be better to encapsulate this & not expose a setter? This obviously depends on how the object is used throughout the rest of the application.

Reply:
While it is bad practice to throw an exception from the `get` it is acceptable throw on `set` – a number of framework classes do this, for example [StringBuilder.Capacity](https://docs.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.capacity?view=netframework-4.8#exceptions).

In many cases, I prefer immutable classes (and the first draft of this class was immutable) but I can't make `EmailAddress` immutable due to its use as configuration (the mutable properties are required for deserialization).

As we're not developing a public library I'm not sure we'd gain much by refactoring here. Even if we were developing a public library I think I'd still be inclined not to refactor (the `Address` set exceptions). Other than the exception occurring slightly sooner, I'm not sure how an exception thrown in a constructor would be any more obvious than one thrown in the set. Either way, it's a '[boneheaded exception](https://ericlippert.com/2008/09/10/vexing-exceptions/)' (Lippert) that's going to occur if a developer tries to set `Address` to `null`.

Reply:
One suggestion would be to separate out the concerns of reading the email address settings and the immutability of the model. We will need to at some point validate all configuration at start up. Here is one resource I found, sure there is plenty more [Configuration validation on startup](https://andrewlock.net/adding-validation-to-strongly-typed-configuration-objects-in-asp-net-core/). By validating the configuration we can ensure the address property is not null & potentially provide a customized error message.

Although I haven't tried it, I believe Json.Net does support deserialization using a constructor [Deserialize using constructor](https://www.newtonsoft.com/json/help/html/DeserializeConstructorHandling.htm). It would mean bringing in another dependency into the project just to support this, which is a little drastic. One to keep an eye on if more requirements come in for this type of process.

As it stands I believe the above change will produce the same result & is the simplest. Just wanted to provide a couple more suggestions.

Reply:
Json.Net does support deserialization using a non-default constructor but the configuration system uses `System.Text.Json` as far as I'm aware (and I don't think it's configurable). I believe it is possible to deserialize an immutable instance using `System.Text.Json` by writing a custom converter.

I considered having a separate class for the configuration but I felt it was going to result in duplication and breaking the DRY principle.

I'm not sure it's worth spending much more time on this now but it's certainly something we could come back to later – it would be good just to get the basic functionality in for now.

Suggestions are always welcome though 👍